### PR TITLE
Master imp seb

### PR DESCRIPTION
**Pull request by seb-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/fields/relational_fields.js [hunk 1]
```diff
@@ -1297,8 +1297,8 @@ var FieldOne2Many = FieldX2Many.extend({
         ev.stopPropagation();
         if (this.editable) {
             if (!this.activeActions.create) {
-                if (ev.data.onFail) {
-                    ev.data.onFail();
+                if (data.onFail) {
```
Comment (seb-odoo) [inline addons/web/static/src/js/fields/relational_fields.js]: where is data defined? another commit?

File: doc/reference/qweb.rst [hunk 2]
```diff
@@ -547,8 +545,8 @@ API
     The QWeb "renderer", handles most of QWeb's logic (loading,
     parsing, compiling and rendering templates).
 
-    OpenERP Web instantiates one for the user in the core module, and 
-    exports it to ``core.qweb``. It also loads all the template files 
+    OpenERP Web instantiates one for the user in the core module, and
```
Comment (pedrobaeza) [inline doc/reference/qweb.rst]: OpenERP?

Comment (seb-odoo) [reply doc/reference/qweb.rst]: Nice catch thanks, I'll update it.

File: addons/web/static/src/js/fields/relational_fields.js (hunk 1 above)
Comment (seb-odoo) [reply addons/web/static/src/js/fields/relational_fields.js]: fixed

**Comment by seb-odoo** [comment]
@ged-odoo could you review this PR please?

File: addons/web/static/src/js/fields/relational_fields.js (hunk 1 above)
Comment (seb-odoo) [reply addons/web/static/src/js/fields/relational_fields.js]: fixed

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 3]
```diff
@@ -1986,7 +1986,9 @@ var BasicModel = AbstractModel.extend({
     },
     _fetchNameGets: function (list, fieldName) {
```
[the remark this answers is not in the document]
Comment (ged-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: list must be a datapoint of type 'list' (for example: a datapoint representing a x2many), and fieldName is the name of a field of type many2one or reference

File: addons/web/static/tests/fields/relational_fields_tests.js [hunk 4]
```diff
@@ -113,6 +113,7 @@ QUnit.module('relational_fields', {
                         ["product", "Product"], ["partner", "Partner"]]},
                     product_id: {string: "Product", type: "many2one", relation: 'product', required: true},
                     partner_ids: {string: "Partner", type: "many2many", relation: 'partner'},
+                    not_required_product_id: {string: "Product", type: "many2one", relation: 'product'},
```
Comment (ged-odoo) [inline addons/web/static/tests/fields/relational_fields_tests.js]: add this field in the test

File: doc/reference/javascript_reference.rst [hunk 5]
```diff
@@ -1481,6 +1481,9 @@ order.
     This field's job is to be displayed as a *handle* in a list view, and allow
     reordering the various records by drag and dropping lines.
 
+    Having more than one field with a handle widget on the same list may lead
+    to unexpected behaviour and should probably not be done.
```
Comment (ged-odoo) [inline doc/reference/javascript_reference.rst]: not supported

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 6]
```diff
@@ -1740,7 +1740,7 @@ var BasicModel = AbstractModel.extend({
                     position: command.position,
                 };
                 def = this._addX2ManyDefaultRecord(list, options).then(function (id) {
-                    if (command.position === 'bottom' && list.orderedResIDs.length >= list.limit) {
+                    if (command.position === 'bottom' && list.orderedResIDs && list.orderedResIDs.length >= list.limit) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: add minitest

File: addons/web/static/tests/fields/relational_fields_tests.js [hunk 7]
```diff
@@ -9491,6 +9491,62 @@ QUnit.module('relational_fields', {
...
+
+        var form = createView({
+            View: FormView,
+            model: 'partner',
+            data: this.data,
+            arch:
+                '<form string="Partners">' +
+                    '<field name="p">' +
+                        '<tree editable="bottom">' +
+                            '<control>' +
```
Comment (ged-odoo) [inline addons/web/static/tests/fields/relational_fields_tests.js]: <controls>
</controls>
<columns>
</columns>

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 8]
```diff
@@ -4280,6 +4298,101 @@ var BasicModel = AbstractModel.extend({
...
+
+    /**
+     * Compute the default value that the handle field should take
+     * We need to compute this in order for new lines to be added at the correct position.
+     *
+     * @param {Object} listID
+     * @param {string} position
+     * @param {string} handleField
+     */
+    _computeOverrideDefaultFields: function (listID, position) {
```
[the remark this answers is not in the document]
Comment (ged-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: alphabetic order

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 9]
```diff
@@ -4280,6 +4298,101 @@ var BasicModel = AbstractModel.extend({
...
+
+        var list = this.localData[listID];
+        var handleField;
+
+        if (! list) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: add comment on this (lists and records

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 10]
```diff
@@ -4280,6 +4298,101 @@ var BasicModel = AbstractModel.extend({
...
+            return false;
+        }
+
+        // We compute the list (get) to apply the pending changes before doing our work,
+        // otherwise new lines might not be taken into account.
+        // We use raw: true because we only need to load the first level of relation.
+        var computedList = this.get(list.id, {raw: true});
+
+        // We don't need to worry about the position of a new line if the list is empty.
+        if (! computedList || ! computedList.data || ! computedList.data.length) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: !computedList

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 11]
```diff
@@ -4280,6 +4298,101 @@ var BasicModel = AbstractModel.extend({
...
+            return false;
+        }
+
+        /*
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: comments

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 12]
```diff
@@ -4280,6 +4298,101 @@ var BasicModel = AbstractModel.extend({
...
+            by 1 (at least until there is a gap in the numbering).
+
+            We don't do it now because it's not an important case.
+        */
+        var handleFieldValue = computedList.data[index].data[handleField];
+        handleFieldValue = position !== 'top' ? handleFieldValue : handleFieldValue - 1;
+
+        var ret = {};
+        ret[handleField] = handleFieldValue;
+        return ret;
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: return {
   handleField: ...,
   value: ...
};

**Comment by seb-odoo** [comment]
@ged-odoo I applied all the recommendations, could you review the PR again please?

**Pull request merged**